File: python/fecg_open/core/ECG_shrinkage0.py
"""
ECG_shrinkage0: Isolate the periodic (beat-aligned) ECG component using SVD-based
optimal shrinkage. Used both to extract the mECG template and to subtract it,
leaving the fECG residual.

Method: Build a beat matrix V (beat-window × beat-count); apply operator-norm optimal
        shrinkage to V's singular values; reconstruct with raised-cosine overlap-add
        weighting at beat boundaries.
        Window half-width = 95th percentile of RRI × 0.5.

Note: OptimalShrinkageOpt='op' (operator norm) is hardcoded; 'fro'/'nuc' branches removed.
"""

import numpy as np
from scipy import stats
from .optimal_shrinkage import optimal_shrinkage


def ECG_shrinkage0(x0, x0_real, current_beats, sigma_coeff, ifauto):
    """
    Optimal SVD shrinkage of a multi-beat ECG matrix.

    Parameters
    ----------
    x0 : ndarray
        Pre-processed ECG signal (1-D row vector)
    x0_real : ndarray
        Original unprocessed ECG signal (1-D row vector, same length as x0)
    current_beats : ndarray
        Sample indices of detected R-peaks (1-based, MATLAB convention)
    sigma_coeff : float
        Noise-level scaling coefficient
    ifauto : int
        If 1, auto-estimate noise; if 0, use fixed sigma=1

    Returns
    -------
    Om0 : ndarray
        Denoised version of x0 (1-D)
    Om0_real : ndarray
        Denoised version of x0_real (1-D)
    """

    x0 = np.asarray(x0).flatten()
    x0_real = np.asarray(x0_real).flatten()
    current_beats = np.asarray(current_beats).flatten().astype(int)

    # Compute beat-window dimensions from RR-interval distribution
    RRI = np.diff(current_beats)
    RRI = np.concatenate(([RRI[0]], RRI))  # prepend first interval

    # Use 95th-percentile RR as proxy for QT interval extent
    # CRITICAL: MATLAB prctile uses (i-0.5)/n cumulative probabilities,
    # numpy percentile uses p/100*(n-1) linear indexing — different results!
    # Must use MATLAB's method for exact match.
    RRI_sorted = np.sort(RRI)
    n_rri = len(RRI_sorted)
    prob = (np.arange(1, n_rri + 1) - 0.5) / n_rri
    pctile_95 = np.interp(0.95, prob, RRI_sorted)
    MaximalQTp = int(np.ceil(pctile_95 * 4 / 8))
    MaximalQTt = int(np.ceil(pctile_95 * 4 / 8))

    # =========================================================================
    # Trim beats that are too close to signal boundaries
    # =========================================================================
    tmp = np.where(current_beats > MaximalQTp)[0]
    current_beats = current_beats[tmp]
    RRI = RRI[tmp]

    tmp = np.where(current_beats + MaximalQTt <= len(x0))[0]
    current_beats = current_beats[tmp]
    RRI = RRI[tmp]

    if len(current_beats) == 0:
        return x0, x0_real

    # =========================================================================
    # Construct beat matrices V (pre-processed) and V2 (original)
    # =========================================================================
    V = []   # columns = individual beat waveforms from x0
    V2 = []  # columns = individual beat waveforms from x0_real
    II = []  # rows    = sample indices for each beat window

    for ii in range(len(current_beats)):
        # Sample-index range centred on this R-peak
        idx_start = int(current_beats[ii] - MaximalQTp - 1)
        idx_end = int(current_beats[ii] + MaximalQTt)

        if idx_start >= 0 and idx_end <= len(x0):
            V.append(x0[idx_start:idx_end])
            V2.append(x0_real[idx_start:idx_end])
            II.append(np.arange(idx_start, idx_end))

    if len(V) == 0:
        return x0, x0_real

    V = np.array(V).T   # shape: (window_size, num_beats)
    V2 = np.array(V2).T
    II = np.array(II)

    n_t, n_theta = V.shape

    # =========================================================================
    # Apply optimal SVD shrinkage to V (pre-processed signal)
    # =========================================================================

    # Noise estimate: sigma_coeff × RMS of the beat matrix after removing the
    # per-row (per-sample) median. This measures beat-to-beat variation rather
    # than absolute amplitude, making sigma adaptive to signal level.
    median_beat = np.median(V, axis=1, keepdims=True)
    sigma = sigma_coeff * np.sqrt(np.sum((V - median_beat)**2) / (n_t * n_theta))

    # Choose orientation so that beta = min(dim)/max(dim) <= 1
    if n_theta > n_t:
        beta0 = n_t / n_theta
        # SVD of normalized matrix
        U, S, Vt = np.linalg.svd(V / (sigma * np.sqrt(n_theta)), full_matrices=False)
        lambdaOS = S

        if ifauto:
            singvals_shrunk = optimal_shrinkage(lambdaOS, beta0)
        else:
            singvals_shrunk = optimal_shrinkage(lambdaOS, beta0, sigma=1)

        XN0 = sigma * np.sqrt(n_theta) * (U @ np.diag(singvals_shrunk) @ Vt[:n_t, :])
    else:
        beta0 = n_theta / n_t
        U, S, Vt = np.linalg.svd(V.T / (sigma * np.sqrt(n_t)), full_matrices=False)
        lambdaOS = S

        if ifauto:
            singvals_shrunk = optimal_shrinkage(lambdaOS, beta0)
        else:
            singvals_shrunk = optimal_shrinkage(lambdaOS, beta0, sigma=1)

        XN0 = (sigma * np.sqrt(n_t) * (U @ np.diag(singvals_shrunk) @ Vt[:n_theta, :])).T

    # =========================================================================
    # Apply optimal SVD shrinkage to V2 (original signal)
    # =========================================================================

    median_beat2 = np.median(V2, axis=1, keepdims=True)
    sigma = sigma_coeff * np.sqrt(np.sum((V2 - median_beat2)**2) / (n_t * n_theta))

    if n_theta > n_t:
        beta0 = n_t / n_theta
        U, S, Vt = np.linalg.svd(V2 / (sigma * np.sqrt(n_theta)), full_matrices=False)
        lambdaOS = S

        if ifauto:
            singvals_shrunk = optimal_shrinkage(lambdaOS, beta0)
        else:
            singvals_shrunk = optimal_shrinkage(lambdaOS, beta0, sigma=1)

        XN02 = sigma * np.sqrt(n_theta) * (U @ np.diag(singvals_shrunk) @ Vt[:n_t, :])
    else:
        beta0 = n_theta / n_t
        U, S, Vt = np.linalg.svd(V2.T / (sigma * np.sqrt(n_t)), full_matrices=False)
        lambdaOS = S

        if ifauto:
            singvals_shrunk = optimal_shrinkage(lambdaOS, beta0)
        else:
            singvals_shrunk = optimal_shrinkage(lambdaOS, beta0, sigma=1)

        XN02 = (sigma * np.sqrt(n_t) * (U @ np.diag(singvals_shrunk) @ Vt[:n_theta, :])).T

    # =========================================================================
    # Overlap-add reconstruction with raised-cosine tapering
    # =========================================================================

    Om0 = np.zeros(len(x0))
    Om0_real = np.zeros(len(x0_real))

    for s in range(len(current_beats)):
        XN_to_add = XN0[:, s]
        XN2_to_add = XN02[:, s]

        # Determine overlap lengths with neighbouring beats
        if s == 0:
            left_overlap = 2
            if s + 1 < len(current_beats):
                right_overlap = len(np.intersect1d(II[s+1, :], II[s, :]))
            else:
                right_overlap = 2
        elif s == len(current_beats) - 1:
            right_overlap = 2
            left_overlap = len(np.intersect1d(II[s-1, :], II[s, :]))
        else:
            left_overlap = len(np.intersect1d(II[s-1, :], II[s, :]))
            right_overlap = len(np.intersect1d(II[s+1, :], II[s, :]))

        # Guard against degenerate overlaps
        if left_overlap <= 1:
            left_overlap = 2
        if right_overlap <= 1:
            right_overlap = 2

        # Raised-cosine (Hann-like) taper over overlap regions
        window_len = MaximalQTp + MaximalQTt + 1
        W = np.ones(window_len)

        # Left taper
        left_taper = np.sin(np.linspace(0, np.pi/2, left_overlap))**2
        W[:left_overlap] = left_taper

        # Right taper (reversed: high to low from start to end of region)
        right_taper = np.sin(np.linspace(0, np.pi/2, right_overlap))**2
        W[-right_overlap:] = right_taper[::-1]

        # Apply taper and accumulate
        Om0_toadd = (W * XN_to_add)
        Om0_real_toadd = (W * XN2_to_add)

        idx_range = II[s, :]
        Om0[idx_range] += Om0_toadd
        Om0_real[idx_range] += Om0_real_toadd

    # =========================================================================
    # Fill any unprocessed samples (never covered by a beat window) with the
    # original signal. Also fills NaN values from numerical edge cases in the
    # SVD/shrinkage path.
    mask_Om0 = (Om0 == 0) | np.isnan(Om0)
    Om0[mask_Om0] = x0[mask_Om0]

    mask_Om0_real = (Om0_real == 0) | np.isnan(Om0_real)
    Om0_real[mask_Om0_real] = x0_real[mask_Om0_real]

    return Om0, Om0_real

File: matlab/fecg_open/core/ECG_shrinkage0.m
% PURPOSE:  Isolate the periodic (beat-aligned) ECG component using SVD-based
%           optimal shrinkage. Used both to extract the mECG template and to
%           subtract it, leaving the fECG residual.
% INPUTS:   x0            — 1×N double, de-trended mixed ECG (for beat timing)
%           x0_real       — 1×N double, morphology-preserving mixed ECG
%           current_beats — 1×K integer, R-peak sample indices
%           sigma_coeff   — scalar, noise-level multiplier (1.5 in main)
%           ifauto        — 0/1 flag; 0 = use sigma=1 (fixed), 1 = estimate from data
% OUTPUTS:  Om0      — 1×N double, reconstructed periodic ECG component (timing version)
%           Om0_real — 1×N double, reconstructed periodic ECG component (morph version)
% METHOD:   Build a beat matrix V (beat-window × beat-count), apply operator-norm
%           optimal shrinkage to V's singular values, reconstruct with overlap-add
%           weighting at beat boundaries. Window half-width = 95th percentile of RRI × 0.5.
function [Om0, Om0_real] = ECG_shrinkage0(x0, x0_real, current_beats, sigma_coeff, ifauto)
    % optimal shrinkage of ECG
    RRI = diff(current_beats);
    RRI = [RRI(1) RRI];
    MaximalQTp = ceil(quantile(RRI, 0.95)*4/8);
    MaximalQTt = ceil(quantile(RRI, 0.95)*4/8);

    tmp = find(current_beats > MaximalQTp);
    current_beats = current_beats(tmp);
    RRI = RRI(tmp);

    tmp = find(current_beats + MaximalQTt <= length(x0));
    current_beats = current_beats(tmp);
    RRI = RRI(tmp);

    V = [];
    V2 = [];
    II = [];
    count = 0;
    for ii = 1:length(current_beats)
        count = count + 1;
        idx = (current_beats(ii)-MaximalQTp) : (current_beats(ii) + MaximalQTt);
        V(:, count) = x0(idx);
        V2(:, count) = x0_real(idx);
        II = [II; idx];
    end

    [n_t, n_theta] = size(V);

    sigma = (V - (median(V, 2)*(ones(1, size(V, 2))))).^2;
    sigma = sigma_coeff*sqrt(sum(sum(sigma)/n_t)/n_theta);
    [n_t, n_theta] = size(V);

    if n_theta > n_t
        beta0 = n_t/n_theta;
        [ae, be, ce] = svd(V./(sigma*sqrt(n_theta)));
        lambdaOSe = diag(be);
        if ifauto
            singvals = optimal_shrinkage(lambdaOSe, beta0);
        else
            singvals = optimal_shrinkage(lambdaOSe, beta0, 1);
        end
        XN0 = sigma*sqrt(n_theta).*(ae*diag(singvals)*ce(:, 1:n_t)');
    else
        beta0 = n_theta/n_t;
        [ae, be, ce] = svd(V'./(sigma*sqrt(n_t)));
        lambdaOSe = diag(be);
        if ifauto
            singvals = optimal_shrinkage(lambdaOSe, beta0);
        else
            singvals = optimal_shrinkage(lambdaOSe, beta0, 1);
        end
        XN0 = sigma*sqrt(n_t).*(ae*diag(singvals)*ce(:, 1:n_theta)')';
    end

    sigma = (V2 - (median(V2, 2)*(ones(1, size(V2, 2))))).^2;
    sigma = sigma_coeff*sqrt(sum(sum(sigma)/n_t)/n_theta);
    if n_theta > n_t
        beta0 = n_t/n_theta;
        [ae, be, ce] = svd(V2./(sigma*sqrt(n_theta)));
        lambdaOSe = diag(be);
        if ifauto
            singvals = optimal_shrinkage(lambdaOSe, beta0);
        else
            singvals = optimal_shrinkage(lambdaOSe, beta0, 1);
        end
        XN02 = sigma*sqrt(n_theta).*(ae*diag(singvals)*ce(:, 1:n_t)');
    else
        beta0 = n_theta/n_t;
        [ae, be, ce] = svd(V2'./(sigma*sqrt(n_t)));
        lambdaOSe = diag(be);
        if ifauto
            singvals = optimal_shrinkage(lambdaOSe, beta0);
        else
            singvals = optimal_shrinkage(lambdaOSe, beta0, 1);
        end
        XN02 = sigma*sqrt(n_t).*(ae*diag(singvals)*ce(:, 1:n_theta)')';
    end

    Om0 = zeros(1, length(x0));
    Om0_real = zeros(1, length(x0_real));

    for s = 1:length(current_beats)

        XN_to_add = XN0(:, s);
        XN2_to_add = XN02(:, s);

        % reconstruction
        if s == 1
            left_overlap = 2;
            right_overlap = length(intersect(II(s+1, :), II(s, :)));
        elseif s == length(current_beats)
            left_overlap = length(intersect(II(s-1, :), II(s, :)));
            right_overlap = 2;
        else
            left_overlap = length(intersect(II(s-1, :), II(s, :)));
            right_overlap = length(intersect(II(s+1, :), II(s, :)));
        end

        if left_overlap <= 1; left_overlap = 2; end
        if right_overlap <= 1; right_overlap = 2; end

        W = ones(MaximalQTp+MaximalQTt+1, 1);
        W(1:left_overlap) = sin(linspace(0, pi/2, left_overlap)).^2;
        W(end:-1:end-right_overlap+1) = sin(linspace(0, pi/2, right_overlap)).^2;

        Om_toadd = (W.*XN_to_add)';
        Om_real_toadd = (W.*XN2_to_add)';

        Om0(II(s, :)) = Om0(II(s, :)) + Om_toadd;
        Om0_real(II(s, :)) = Om0_real(II(s, :)) + Om_real_toadd;
    end

    Om0(Om0 == 0) = x0(Om0 == 0);
    Om0_real(Om0_real == 0) = x0_real(Om0_real == 0);

end
